### PR TITLE
[SPARK-33504][CORE] The application log in the Spark history server contains sensitive attributes should be redacted

### PR DESCRIPTION
**Pull request by echohlne**
### What changes were proposed in this pull request?
To make sure the sensitive attributes to be redacted in the history server log.

### Why are the changes needed?
We found the secure attributes like password  in SparkListenerJobStart and SparkListenerStageSubmitted events would not been redated, resulting in sensitive attributes can be viewd directly.
The screenshot can be viewed in the attachment of JIRA spark-33504
### Does this PR introduce _any_ user-facing change?
no

### How was this patch tested?
muntual test works well, I have also added unit testcase.

### CODE REVIEW

**Comment by mridulm** [comment]
redaction support looks kind of weird.
For `ApplicationEnvironmentInfo`, the event file contains unredacted event with the api redacting on demand.
On other hand, `SparkListenerEnvironmentUpdate` is redacted before getting logged into event file.

The current pr takes the approach of `SparkListenerEnvironmentUpdate` - and logs redacted events.
Personally I am inclined to have source of truth in the event file, and expose redacted view.

But then, all of these was last changed 4 years or so back (#15971).
Any thoughts @vanzin, @tgravescs ?

**Comment by mridulm** [comment]
ok to test

File: core/src/test/scala/org/apache/spark/scheduler/EventLoggingListenerSuite.scala [hunk 1]
```diff
@@ -97,6 +97,17 @@ class EventLoggingListenerSuite extends SparkFunSuite with LocalSparkContext wit
...
   }
 
+  test("Spark-33504 password redaction in properties") {
+    val key = "spark.executorEnv.HADOOP_CREDSTORE_PASSWORD"
+    val secretPassword = "secret_password"
+    val conf = getLoggingConf(testDirPath, None).set(key, secretPassword)
+    val properties = new Properties()
+    properties.setProperty(key, secretPassword)
+    val eventLogger = new EventLoggingListener("test", None, testDirPath.toUri(), conf)
+    val redactedProperties = eventLogger.redactProperties(properties)
```
Comment (mridulm) [inline core/src/test/scala/org/apache/spark/scheduler/EventLoggingListenerSuite.scala:107]: Test whether `SparkListenerJobStart` is updated as required ? `redactProperties` can be made private

**Comment by akiyamaneko** [comment]
@mridulm done this, please help to review again, thanks!

File: core/src/main/scala/org/apache/spark/scheduler/EventLoggingListener.scala [hunk 2]
```diff
@@ -103,7 +105,7 @@ private[spark] class EventLoggingListener(
 
   // Events that do not trigger a flush
   override def onStageSubmitted(event: SparkListenerStageSubmitted): Unit = {
-    logEvent(event)
+    logEvent(SparkListenerStageSubmitted(event.stageInfo, redactProperties(event.properties)))
```
Comment (mridulm) [inline core/src/main/scala/org/apache/spark/scheduler/EventLoggingListener.scala:108]: Use copy instead - here and elsewhere, to be forward compatible.
For example: `event.copy(properties =  redactProperties(event.properties))`

**Review by mridulm — COMMENTED** [review]
Thanks for fixing the test @akiyamaneko !
I will wait for response from others. I like the idea of the fix itself, and agree it should be redacted - I am trying to understand the history of redaction to make sure we are going the right way.

**Comment by akiyamaneko** [comment]
@mridulm thanks for your great advice, Please review it again when you have time.

**Comment by tgravescs** [comment]
So both of the configs in the jira look like they are user generated configs.   It sounds like you updated the spark.redaction.regex config but that didn't work for these events. It seems like we just missed these, so thanks for catching this.

I think the current approach makes sense and matches our previous implementations.

**Comment by mridulm** [comment]
@tgravescs The question I had was regarding when redaction is applied - while logging the event or when surfacing in UI.
Only for the case of `EventLoggingListener.onEnvironmentUpdate` we are redacting event file itself - in all other cases of application of redaction regex, we do not modify the event file, but rather change when presenting to user.

I am not sure if this was done intentionally (allowing superuser to look at the unredact properties - and so `onEnvironmentUpdate` is a bug) or was an oversight (in which case all properties should be redacted before logging).

Thoughts ? Thanks.

**Comment by tgravescs** [comment]
sorry for my delay, behind on reviews.

I think we should redact from both ui/rest or from event logging as people may not realize event logging has the password and history server could allow viewing from people who shouldn't see those.  If people want it differently I would say we add a config to allow it but have it off by default - but I would wait until someone requests this.   

I think it was an oversight that these events weren't redacted. so if you see other cases we aren't we should fix them.

**Comment by mridulm** [comment]
Thanks for clarifying Tom ! Sounds good to me.
I am fine with this patch given @tgravescs's comments above.

We can have follow up work to redact from other events as well - at event logging time.

**Review by mridulm — APPROVED** [review]

**Comment by akiyamaneko** [comment]
@mridulm @tgravescs  thanks for your review!

**Comment by akiyamaneko** [comment]
@mridulm  Does it  need someone else to continue reviewing, or can someone help me merge the code?

**Comment by tgravescs** [comment]
I will merge this shortly

**Comment by tgravescs** [comment]
thanks @akiyamaneko @mridulm merged to master and branch-3.0

**Comment by dongjoon-hyun** [comment]
Hi, @tgravescs .
There is a compilation error in `branch-3.0`.
```
[error] /home/runner/work/spark/spark/core/src/test/scala/org/apache/spark/scheduler/EventLoggingListenerSuite.scala:118: not found: value resourceProfileId
[error] Error occurred in an application involving default arguments.
[error]       resourceProfileId = ResourceProfile.DEFAULT_RESOURCE_PROFILE_ID)
[error]       ^
```

**Comment by dongjoon-hyun** [comment]
Could you take a look, please? I guess we need to recover the branch-3.0 first by reverting it and make a PR to `branch-3.0` to pass all CIs. But, it's up to you.

**Comment by tgravescs** [comment]
yeah lets revert first

**Comment by tgravescs** [comment]
https://github.com/apache/spark/pull/30576 is revert

@akiyamaneko could you port this to branch-3.0 and put up a PR?

**Comment by tgravescs** [comment]
@akiyamaneko  do you have time to back port?

**Comment by akiyamaneko** [comment]
@tgravescs  sorry for delay, Is there anything I can do？

**Comment by tgravescs** [comment]
I don't think anyone back ported it yet so if you have time it would be great to put up a version again branch-3.0

**Pull request closed without merging**